src/experimental/theoretic_information.cpp

```diff
@@ -0,0 +1,644 @@
+
+        cell_limits[2] = limit_y;
+        cell_limits[3] = limit_y + m_resolution;
+    }
+
+    if (final_grid_pos[0] >= robot_grid_pos[0] && final_grid_pos[1] < robot_grid_pos[1])
+    {
+        // X greater and Y minor. WRO final points
+        initial_y[2] = limit_y - m_resolution;
+        initial_y[3] = limit_y - m_resolution;
+
+        final_y[1] = limit_y - m_resolution;
+        final_y[3] = limit_y - m_resolution;
+
+        cell_limits[2] = limit_y - m_resolution;
+        cell_limits[3] = limit_y;
+    }
+
+    if (final_grid_pos[0] < robot_grid_pos[0] && final_grid_pos[1] < robot_grid_pos[1])
+    {
+        // X minor and Y minor. WRO final points
+        initial_x[2] = limit_x - m_resolution;
+        initial_x[3] = limit_x - m_resolution;
+        initial_y[2] = limit_y - m_resolution;
+        initial_y[3] = limit_y - m_resolution;
+
+        final_x[0] = limit_x - m_resolution;
+        final_x[2] = limit_x - m_resolution;
+        final_y[1] = limit_y - m_resolution;
+        final_y[3] = limit_y - m_resolution;
+
+        cell_limits[0] = limit_x - m_resolution;
+        cell_limits[1] = limit_x;
+        cell_limits[2] = limit_y - m_resolution;
+        cell_limits[3] = limit_y;
+    }
+}
+
+float TeorethicInformation::calculateLogs(float probability)
+{
+    /*
+        To calculate the log-odds
+    */
+    return log(probability / (1 - probability));
+}
+
+float TeorethicInformation::calculateMapMutualInformation()
+{
+    /*
+        To calculate map mutual information, this is the summation
+        of all cells mutual information
+    */
+    float sum = 0.0f;
+    for (int i = 0; i < m_num_cells; ++i)
+    {
+        for (int j = 0; j < m_num_cells; ++j)
+        {
+            sum += m_mutual_grid[i][j];
+        }
+    }
+    return sum;
+}
+
+std::unordered_map<TeorethicInformation::map_tuple, float, TeorethicInformation::HashTuple> TeorethicInformation::computeProbabilities(std::vector<std::vector<float>>& meas_outcm)
+{
+    /*
+        To compute all the possible combinations of a grid cell, given a set of measurement outcomes
+    */
+    // Cleaning measurement outcomes map
+    std::unordered_map<map_tuple, float, HashTuple> map_out;
+    std::unordered_map<map_tuple, float, HashTuple>::iterator it_out;
+
+    // The number of measurements
+    int k = meas_outcm.size(); 
+    int r = 1;
+
+    float p_free = meas_outcm[0][2];
+    float p_occ = meas_outcm[0][1];
+    float p_un = meas_outcm[0][0];
+
+    // Root
+    map_out.insert(map_pair(std::make_tuple(0, 0, 0), 1.0f));
+
+    // First measurement
+    map_out.insert(map_pair(std::make_tuple(1, 0, 0), p_free));
+    map_out.insert(map_pair(std::make_tuple(0, 1, 0), p_occ));
+    map_out.insert(map_pair(std::make_tuple(0, 0, 1), p_un));
+
+    for (int i = r; r < k; ++r)
+    {
+        std::vector<map_tuple> tup_vct;
+        std::vector<float> acc_prob;
+
+        for (it_out = map_out.begin(); it_out != map_out.end(); ++it_out)
```

Comment:
@kmilo7204 nit: in C++11 and onwards, `range-for` and `auto` are your friends:
```suggestion
        for (auto & pair : map_out)
```
You can then do `pair.first` and `pair.second` to fetch key and value respectively.